### ISSUE TITLE
SYNCH_WAIT cause "module task timeout"

### ISSUE THREAD

**Issue by kinstaky** [xiallc/PixieSDK#2]
Hardware:
  crate: type UEP 6021
  module: 
    slot2: 500M 14bits revF, firmware version 43351, CSRB = 65
    slot3: 100M 14bits revF, firmware version 40391, CSRB = 0
    slot4: 250M 14bits revF, firmware version 44496, CSRB = 0

Description:
I tried to run the example example_pixie16api at the git commit 27ba537b4acb87a1911af9514a89f092b33b50d8 but failed. I use command `./example_pixie16api list-mode -t 1 --synch-wait=1 --num-runs=1 --config config.json` to run in list mode for 1 seconds with the modules start/stop synchronously. But I got the error "Pixie16EndRun failed with code -214 and message: module task timeout". But it works with option `--synch-wait=0`.
And these modules also work in the old version legacy api. So I tried to compare the difference between them.
In the legacy code, line 1299 of file pixie16app.c gives `(void) Pixie_Register_IO(modulenumber, PCI_STOPRUN_RGADDR, MOD_WRITE, &dummy);`. It seems that it write an uninitialized variable `dummy` to the virtual memory at address 0x44. But I didn't find anything like this in the new version code.
And in the new version code, I found the example in my case always failed at the check in line 164 in function "void end(module::module &module)" in file run.cpp.  The condition check is `(!hw::run::active(module) || run_ended(module))` and modules failed to pass it. For the active check, although the RUNENA(1) bit was clear, the RUNACTIVE(13) bit didn't clear. And I'm not sure whether there may be some problem in the second `run_ended` check. This check at first require run task to be histogram or list mode (at line 117,118); but run task was set to run_stopping at line 156 before the check. So I think no module could pass this check. And then it require the module is not the leader module. I think it means that in my case, module in slot 3 and 4 with CSRB==0 could pass the check, but actually they didn't because the run task invalid.
Then I tried to add something like `Pixie_Register_IO(0,0x44,MOD_WRITE,&dummy)` from legacy code to new version code. I tried to add `module->write_word(0x44, 0);` at line 817 in pixie/pixie16/pixie16.cpp and comment line 156 `module.run_task = ruN_task::run_stopping` in file run.cpp. As expected, it doesn't work. But it clears the RUNACTIVE(13) bit of module 0. So I add `module->write_word(0x44,0)` at line 813 in pixie/pixie16/pixie16.cpp and change line 435 in example_pixie16api.cpp from `cfg.module[0].number` to `cfg.num_modules()`. Then it works and I got successful output like "INFO - Pixie16ExitSystem finished successfully.". But I didn't check the acquiring data yet.
I hope to know why making synch-wait=1 cause error. Maybe because I didn't set the parameters properly? And is it reasonable to write something to address 0x44? Whether this operation keep the modules start and stop synchronously? If not, how to make it right?

**Comment by kinstaky** [comment]
I found that I made some mistakes. Now I find how to trigger the error (code 214) stably.
 + boot all modules (in my case is 3 modules)
 + edit the config file, and delete some modules (in my case I delete the last module)
 + run the residual modules (in my case I ran modules in slot 2 and 3)
Then I get error with code 214 and message "module task timeout".

**Comment by xia-stan** [comment]
@kinstaky : Sorry for the delay responding. Running a synchronous run requires that all modules participate. If you have 3 modules in the system you have to run a synchronous run with all 3. One module will need to be defined as the leader module. This is usually `MOD_CSRB = 65`. 

I'm working on a FAQ entry related to synchronous data runs.

**Comment by kinstaky** [comment]
I know that I should run all 3 modules  if I booted all of them. But what will happen if I just booted 2 of them? I have booted the modules for twice. The first time I booted 3 modules but the second time I just booted 2 of them. I think after the second boot, PixieSDK may regard the 2 booted modules as "all" modules. So I can just run 2 modules in synchronous mode, since they are "all" modules the software found. I guess the software defining "all" modules as all modules booted this time. But actually the hardware or something have some memory about the last boot. I hope that I have described my question clearly.

**Comment by xia-stan** [comment]
@kinstaky : I apologize for not clarifying the synchronous runs better. Let me first define a few terms for the discussion below. 

---
- **physical module** : A module that's been installed in the Pixie-16 Crate.
- **booted module** : A physical module that's been made ready for use by loading the firmware. This firmware persists in the module until the power is cycled or another boot occurs. 
- **leader module** : A booted module that's been defined as the list-mode leader during a synchronous data run.
- **virtual module**: A software object that maps a physical module into the SDK.
- **online module** : A virtual module the SDK identified that has been successfully booted and made ready for use.
- **offline module**: A virtual module the SDK identified, but could not make ready to use or the user didn't request use of.
---

The SDK finds all physical modules in the system during initialization. The user may use any combination of physical modules in the system. The SDK matches this list of physical modules against the user's requested modules. The SDK will consider "all" modules to be the total number of online modules. 

Synchronous list-mode runs execute through a combination of hardware, firmware and software. First, the SDK verifies that all modules have the correct settings for a synchronous list-mode data run. This check happens before the user requests a run. When the user requests a synchronized list-mode data run, the leader module's DSP checks the status of a backplane line in the crate. The run starts when all physical modules set this line to the same state. If, for any reason, a physical module cannot set this line, the data run will never start. 

In short, all physical modules in the system must participate in a synchronous data run, or none of them can participate.
